A39-xds-http-filters.md

```diff
@@ -0,0 +1,152 @@
+  resource to be NACKed.
+- Every filter in the list must have a unique instance name.  Any
+  duplicate instance names will cause the Listener resource to be NACKed.
+
+Note that gRPC will support configuring multiple instances of the same
+filter implementation, each with its own name and config.
+
+In each of the `VirtualHost`, `Route`, and `WeightedCluster.ClusterWeight`
+protos, there is a field called `typed_per_filter_config`, which is a map
+from filter instance name to a corresponding filter-specific configuration
+proto message.  (Note that Envoy currently keys these maps by legacy
+filter names instead of by filter instance names, but this is expected
+to change, as per https://github.com/envoyproxy/envoy/issues/12274; gRPC
+will directly implement the desired semantics here.)
+
+When gRPC receives the `RouteConfiguration` proto, either in the LDS or
+RDS response, it will validate the contents of each
+`typed_per_filter_config` map as follows:
+- Every config in the list must be of a known type, as identified by the
+  map value proto message type.  Any unknown config type will cause the
+  resource to be NACKed.
+- Note that gRPC will *not* fail validation if the map key specifies a
+  filter instance name that does not exist in the `HttpConnectionManager`
+  filter list.  This is because during an update, the xDS client code
+  cannot know which `HttpConnectionManager` config is currently being
+  used.
+
+Unlike Envoy, which provides an API that filters must call to access their
+per-{VirtualHost,Route,ClusterWeight} config overrides, gRPC will
+require filters to construct a merged configuration, applying any
+necessary overrides, when the xDS config is applied.  The filters will
+then use this merged configuration at run-time.  However, this approach
+is an implementation detail that could change in the future.
+
+### Initially Supported Filter: Router
+
+Initially, gRPC will support only one filter, the [router
+filter](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/router_filter),
+which is used for the proto message type
+[`extensions.filters.http.router.v3.Router`](https://github.com/envoyproxy/envoy/blob/18db4c90e3295fb2c39bfc7b2ce641cfd6c3fbed/api/envoy/extensions/filters/http/router/v3/router.proto#L23).
+As in Envoy, this filter is required and must be last in the filter list.
+
+Note that gRPC will not actually have a discrete filter implementation
+for the router filter; instead, this filter will simply trigger the
+built-in routing functionality that we have already implemented in the
+resolver and LB policy plugins to the gRPC client channel.  However,
+in order to retain compatibility with Envoy, gRPC will still accept
+this filter configuration and treat it similarly to the way that Envoy
+does.
+
+Specifically, if the router filter is not present, gRPC will fail all
+requests.  In addition, any filters in the list after the router filter
```

Comment:
Is this a breaking change assuming the current gRPC xDS implementation does not look for the router filter?

Reply:
Yes, it could be, which is why it's called out here.

If the xDS server is not currently populating any HTTP filters, upgrading gRPC without changing the xDS server to populate the router filter will cause failures.  And if the xDS server is populating HTTP filters meant for Envoy that gRPC does not support (which the current version of gRPC is ignoring), upgrading gRPC without changing the xDS server to populate only the router filter will cause failures.  So the right upgrade path is to first change the xDS server to populate only the router filter (or the router filter and maybe the fault injection filter, since we'll be releasing both features at the same time), and then upgrade their clients.

Adding support for HTTP filters isn't really fundamentally different than adding support for any new xDS field.  Whenever we start to support a new feature, existing clients may break.  For example, when we add circuit breaking support, the default for that is a max of 1024 concurrent requests, so an existing client that is busier than that will start breaking.  Unfortunately, I think this is just something people need to be used to with xDS.

Reply:
Understood but the consequences are quite drastic "...if the router filter is not present, gRPC will fail all
requests...".

At the same time most current users of gRPC with xDS probably use Envoy-conforming xDS config which always has the router filter so in practice won't break anybody.

Reply:
Yeah, I don't think this is likely to affect too many people.  There may be some that are not sending HTTP filter configs at all, but it shouldn't be too hard for them to identify and fix the problem.